InvalidDepositEventsException on restart

Getting the exception below on restart across a number of boxes including seen locally once.

First problem is that it's only being logged at debug level and should be a big clear error message on the console.

Second problem is it doesn't resolve and we lose track of eth1 and start voting poorly.  Restarting (possibly after enough time passes) solves the issue.

We seem to be calculating the first block to request incorrectly after startup.

```
{"timestamp":"2021-07-07T04:57:17,851","level":"DEBUG","thread":"pool-17-thread-201","class":"DepositFetcher","message":"Failed to request deposit events for block numbers in the range (5049706, 5049706). Retrying.","throwable":" java.util.concurrent.CompletionException: tech.pegasys.teku.ethereum.pow.api.InvalidDepositEventsException: Expected next deposit
at index 128373, but got 128372
at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1155)
at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315)
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:529)
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:32)
at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:142)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$27(SafeFuture.java:367)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315)
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:529)
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:32)
at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:142)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$27(SafeFuture.java:367)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315)
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:529)
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:32)
at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:142)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$27(SafeFuture.java:367)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315)
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:529)
at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:32)
at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:142)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$27(SafeFuture.java:367)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137)
at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:147)
at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137)
at org.web3j.utils.Async.lambda$run$1(Async.java:38)
at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1800)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
at java.base/java.lang.Thread.run(Thread.java:832)\nCaused by: tech.pegasys.teku.ethereum.pow.api.InvalidDepositEventsException: Expected next deposit
at index 128373, but got 128372
at tech.pegasys.teku.ethereum.pow.api.InvalidDepositEventsException.expectedDepositAtIndex(InvalidDepositEventsException.java:26)
at tech.pegasys.teku.pow.ValidatingEth1EventsPublisher.assertDepositEventIsValid(ValidatingEth1EventsPublisher.java:56)
at tech.pegasys.teku.pow.ValidatingEth1EventsPublisher.onDepositsFromBlock(ValidatingEth1EventsPublisher.java:46)
at tech.pegasys.teku.pow.DepositFetcher.postDeposits(DepositFetcher.java:220)
at tech.pegasys.teku.pow.DepositFetcher.postEventsForBlock(DepositFetcher.java:186)
at tech.pegasys.teku.pow.DepositFetcher.postDepositEvents(DepositFetcher.java:157)
at tech.pegasys.teku.pow.DepositFetcher.lambda$postDepositEvents$4(DepositFetcher.java:173)
at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1146)\n\t... 70 more\n"}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,7 @@ For information on changes in released versions of Teku, see the [releases page]
 - Fix an issue where new peers would not be found after a network outage.
 - Fix a file handle leak in jvm-libp2p.
 - `/eth/v1/beacon/pool/sync_committees` incorrectly returned 503 when there were no errors instead of 200.
+- Fix an issue where deposits for the PoW chain could be loaded out of order on restart.
 
 
 ### Experimental: New Altair REST APIs

--- a/pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java
+++ b/pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java
@@ -42,6 +42,7 @@ public class Eth1DepositManager {
   private final DepositProcessingController depositProcessingController;
   private final MinimumGenesisTimeBlockFinder minimumGenesisTimeBlockFinder;
   private final Optional<UInt64> depositContractDeployBlock;
+  private final Eth1HeadTracker headTracker;
 
   public Eth1DepositManager(
       final Eth1Provider eth1Provider,
@@ -50,7 +51,8 @@ public class Eth1DepositManager {
       final Eth1DepositStorageChannel eth1DepositStorageChannel,
       final DepositProcessingController depositProcessingController,
       final MinimumGenesisTimeBlockFinder minimumGenesisTimeBlockFinder,
-      final Optional<UInt64> depositContractDeployBlock) {
+      final Optional<UInt64> depositContractDeployBlock,
+      final Eth1HeadTracker headTracker) {
     this.eth1Provider = eth1Provider;
     this.asyncRunner = asyncRunner;
     this.eth1EventsPublisher = eth1EventsPublisher;
@@ -58,6 +60,7 @@ public class Eth1DepositManager {
     this.depositProcessingController = depositProcessingController;
     this.minimumGenesisTimeBlockFinder = minimumGenesisTimeBlockFinder;
     this.depositContractDeployBlock = depositContractDeployBlock;
+    this.headTracker = headTracker;
   }
 
   public void start() {
@@ -165,12 +168,11 @@ public class Eth1DepositManager {
   }
 
   private SafeFuture<EthBlock.Block> getHead() {
-    return eth1Provider
-        .getGuaranteedLatestEth1Block()
-        .thenApply(EthBlock.Block::getNumber)
-        .thenApply(UInt64::valueOf)
-        .thenCompose(this::waitForEth1ChainToReachFollowDistanceIfNecessary)
-        .thenApply(number -> number.minus(Constants.ETH1_FOLLOW_DISTANCE))
+    final SafeFuture<UInt64> headBlockNumber = new SafeFuture<>();
+    final long subscriberId = headTracker.subscribe(headBlockNumber::complete);
+
+    return headBlockNumber
+        .thenPeek(__ -> headTracker.unsubscribe(subscriberId))
         .thenCompose(eth1Provider::getGuaranteedEth1Block)
         .exceptionallyCompose(
             (err) -> {
@@ -180,28 +182,9 @@ public class Eth1DepositManager {
                   Constants.ETH1_DEPOSIT_REQUEST_RETRY_TIMEOUT.toSeconds(),
                   err);
 
-              return asyncRunner
-                  .getDelayedFuture(Constants.ETH1_DEPOSIT_REQUEST_RETRY_TIMEOUT)
-                  .thenCompose((__) -> getHead());
+              return asyncRunner.runAfterDelay(
+                  this::getHead, Constants.ETH1_DEPOSIT_REQUEST_RETRY_TIMEOUT);
             });
-  }
-
-  private SafeFuture<UInt64> getLatestEth1BlockNumber() {
-    return eth1Provider
-        .getLatestEth1Block()
-        .thenApply(EthBlock.Block::getNumber)
-        .thenApply(UInt64::valueOf);
-  }
-
-  private SafeFuture<UInt64> waitForEth1ChainToReachFollowDistanceIfNecessary(UInt64 number) {
-    if (number.isLessThan(Constants.ETH1_FOLLOW_DISTANCE)) {
-      return asyncRunner
-          .getDelayedFuture(Constants.ETH1_LOCAL_CHAIN_BEHIND_FOLLOW_DISTANCE_WAIT)
-          .thenCompose(__ -> getLatestEth1BlockNumber())
-          .thenCompose(this::waitForEth1ChainToReachFollowDistanceIfNecessary);
-    } else {
-      return SafeFuture.completedFuture(number);
-    }
   }
 
   private BigInteger getFirstUnprocessedBlockNumber(

--- a/pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java
+++ b/pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java
@@ -168,10 +168,10 @@ public class Eth1DepositManager {
   }
 
   private SafeFuture<EthBlock.Block> getHead() {
-    final SafeFuture<UInt64> headBlockNumber = new SafeFuture<>();
-    final long subscriberId = headTracker.subscribe(headBlockNumber::complete);
+    final SafeFuture<UInt64> headBlockNumberFuture = new SafeFuture<>();
+    final long subscriberId = headTracker.subscribe(headBlockNumberFuture::complete);
 
-    return headBlockNumber
+    return headBlockNumberFuture
         .thenPeek(__ -> headTracker.unsubscribe(subscriberId))
         .thenCompose(eth1Provider::getGuaranteedEth1Block)
         .exceptionallyCompose(

--- a/pow/src/test/java/tech/pegasys/teku/pow/Eth1DepositManagerTest.java
+++ b/pow/src/test/java/tech/pegasys/teku/pow/Eth1DepositManagerTest.java
@@ -15,10 +15,10 @@ package tech.pegasys.teku.pow;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.atLeastOnce;
 import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.never;
-import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 import static tech.pegasys.teku.infrastructure.async.SafeFuture.COMPLETE;
@@ -30,6 +30,7 @@ import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
 import org.mockito.InOrder;
 import org.web3j.protocol.core.methods.response.EthBlock.Block;
 import tech.pegasys.teku.ethereum.pow.api.MinGenesisTimeBlockEvent;
@@ -37,6 +38,7 @@ import tech.pegasys.teku.infrastructure.async.SafeFuture;
 import tech.pegasys.teku.infrastructure.async.StubAsyncRunner;
 import tech.pegasys.teku.infrastructure.async.TrackingUncaughtExceptionHandler;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.pow.Eth1HeadTracker.HeadUpdatedSubscriber;
 import tech.pegasys.teku.service.serviceutils.FatalServiceFailureException;
 import tech.pegasys.teku.storage.api.Eth1DepositStorageChannel;
 import tech.pegasys.teku.storage.api.schema.ReplayDepositsResult;
@@ -58,6 +60,7 @@ class Eth1DepositManagerTest {
       mock(DepositProcessingController.class);
   private final MinimumGenesisTimeBlockFinder minimumGenesisTimeBlockFinder =
       mock(MinimumGenesisTimeBlockFinder.class);
+  private final Eth1HeadTracker eth1HeadTracker = mock(Eth1HeadTracker.class);
   private final TrackingUncaughtExceptionHandler exceptionHandler =
       new TrackingUncaughtExceptionHandler();
 
@@ -76,7 +79,8 @@ class Eth1DepositManagerTest {
           eth1DepositStorageChannel,
           depositProcessingController,
           minimumGenesisTimeBlockFinder,
-          Optional.empty());
+          Optional.empty(),
+          eth1HeadTracker);
 
   @BeforeAll
   static void setConstants() {
@@ -97,10 +101,11 @@ class Eth1DepositManagerTest {
   void shouldStartWithNoStoredDepositsAndHeadBeforeMinGenesisTime() {
     final BigInteger headBlockNumber = BigInteger.valueOf(100);
     when(eth1DepositStorageChannel.replayDepositEvents()).thenReturn(NOTHING_REPLAYED);
-    withFollowDistanceHead(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP - 1);
     when(depositProcessingController.fetchDepositsInRange(any(), any())).thenReturn(COMPLETE);
 
     manager.start();
+
+    notifyHeadBlock(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP - 1);
 
     verify(eth1EventsChannel, never()).setLatestPublishedDeposit(any());
     inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
@@ -122,23 +127,24 @@ class Eth1DepositManagerTest {
 
     final BigInteger headBlockNumber = BigInteger.valueOf(100);
     when(eth1DepositStorageChannel.replayDepositEvents()).thenReturn(NOTHING_REPLAYED);
-    // simulate major issue on getGuaranteedLatestEth1Block
-    when(eth1Provider.getGuaranteedLatestEth1Block())
-        .thenReturn(SafeFuture.failedFuture(new IllegalStateException("Unknown Error")));
     when(depositProcessingController.fetchDepositsInRange(any(), any())).thenReturn(COMPLETE);
 
     manager.start();
 
+    // Notify head block but retrieving it will fail
+    notifyHeadBlockAndFailToRetrieve(headBlockNumber);
+
     // Set up initial request to eth1 node to fail
     // We should retry until it succeeds
     for (int i = 0; i < retryCount; i++) {
-      assertThat(asyncRunner.countDelayedActions()).isEqualTo(1);
+      assertThat(asyncRunner.countDelayedActions()).describedAs("on attempt " + i).isEqualTo(1);
       asyncRunner.executeQueuedActions();
+      notifyHeadBlockAndFailToRetrieve(headBlockNumber);
     }
-    // Set up next getHead request to succeed
-    withFollowDistanceHead(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP - 1);
     assertThat(asyncRunner.countDelayedActions()).isEqualTo(1);
+    // Next getHead request succeeds
     asyncRunner.executeQueuedActions();
+    notifyHeadBlock(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP - 1);
 
     inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
     // Process blocks up to the current chain head
@@ -153,16 +159,24 @@ class Eth1DepositManagerTest {
     assertNoUncaughtExceptions();
   }
 
+  private void notifyHeadBlockAndFailToRetrieve(final BigInteger headBlockNumber) {
+    when(eth1Provider.getGuaranteedEth1Block(any(UInt64.class)))
+        .thenReturn(SafeFuture.failedFuture(new IllegalStateException("Unknown Error")));
+    final ArgumentCaptor<HeadUpdatedSubscriber> captor =
+        ArgumentCaptor.forClass(HeadUpdatedSubscriber.class);
+    verify(eth1HeadTracker, atLeastOnce()).subscribe(captor.capture());
+    captor.getValue().onHeadUpdated(UInt64.valueOf(headBlockNumber));
+  }
+
   @Test
   void shouldFailIfStorageReplayFails() {
-    final BigInteger headBlockNumber = BigInteger.valueOf(100);
     when(eth1DepositStorageChannel.replayDepositEvents())
         .thenReturn(SafeFuture.failedFuture(new IllegalStateException("Fail")));
-    withFollowDistanceHead(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP - 1);
     when(depositProcessingController.fetchDepositsInRange(any(), any())).thenReturn(COMPLETE);
 
     manager.start();
 
+    verify(eth1HeadTracker, never()).subscribe(any());
     inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
     inOrder.verifyNoMoreInteractions();
     assertThat(exceptionHandler.getUncaughtExceptions()).hasSize(1);
@@ -179,10 +193,10 @@ class Eth1DepositManagerTest {
         .thenReturn(
             SafeFuture.completedFuture(
                 ReplayDepositsResult.create(lastReplayedBlock, lastReplayedDepositIndex, false)));
-    withFollowDistanceHead(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP - 1);
     when(depositProcessingController.fetchDepositsInRange(any(), any())).thenReturn(COMPLETE);
 
     manager.start();
+    notifyHeadBlock(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP - 1);
 
     verify(eth1EventsChannel).setLatestPublishedDeposit(UInt64.valueOf(lastReplayedDepositIndex));
     inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
@@ -210,11 +224,11 @@ class Eth1DepositManagerTest {
         .thenReturn(
             SafeFuture.completedFuture(
                 ReplayDepositsResult.create(lastReplayedBlock, lastReplayedDepositIndex, false)));
-    withFollowDistanceHead(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
     withMinGenesisBlock(headBlockNumber, minGenesisBlockNumber);
     when(depositProcessingController.fetchDepositsInRange(any(), any())).thenReturn(COMPLETE);
 
     manager.start();
+    notifyHeadBlock(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
 
     verify(eth1EventsChannel).setLatestPublishedDeposit(UInt64.valueOf(lastReplayedDepositIndex));
     inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
@@ -250,11 +264,11 @@ class Eth1DepositManagerTest {
     final BigInteger headBlockNumber = BigInteger.valueOf(100);
     final BigInteger minGenesisBlockNumber = BigInteger.valueOf(60);
     when(eth1DepositStorageChannel.replayDepositEvents()).thenReturn(NOTHING_REPLAYED);
-    withFollowDistanceHead(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
     withMinGenesisBlock(headBlockNumber, minGenesisBlockNumber);
     when(depositProcessingController.fetchDepositsInRange(any(), any())).thenReturn(COMPLETE);
 
     manager.start();
+    notifyHeadBlock(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
 
     inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
     // Find min genesis block
@@ -293,10 +307,10 @@ class Eth1DepositManagerTest {
         .thenReturn(
             SafeFuture.completedFuture(
                 ReplayDepositsResult.create(lastReplayedBlock, lastReplayedDepositIndex, true)));
-    withFollowDistanceHead(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
     when(depositProcessingController.fetchDepositsInRange(any(), any())).thenReturn(COMPLETE);
 
     manager.start();
+    notifyHeadBlock(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
 
     verify(eth1EventsChannel).setLatestPublishedDeposit(UInt64.valueOf(lastReplayedDepositIndex));
     inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
@@ -318,10 +332,10 @@ class Eth1DepositManagerTest {
         .thenReturn(
             SafeFuture.completedFuture(
                 ReplayDepositsResult.create(lastReplayedBlock, lastReplayedDepositIndex, true)));
-    withFollowDistanceHead(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
     when(depositProcessingController.fetchDepositsInRange(any(), any())).thenReturn(COMPLETE);
 
     manager.start();
+    notifyHeadBlock(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
 
     verify(eth1EventsChannel).setLatestPublishedDeposit(UInt64.valueOf(lastReplayedDepositIndex));
     inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
@@ -343,10 +357,10 @@ class Eth1DepositManagerTest {
         .thenReturn(
             SafeFuture.completedFuture(
                 ReplayDepositsResult.create(lastReplayedBlock, lastReplayedDepositIndex, false)));
-    withFollowDistanceHead(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
     when(depositProcessingController.fetchDepositsInRange(any(), any())).thenReturn(COMPLETE);
 
     manager.start();
+    notifyHeadBlock(headBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP + 1000);
 
     verify(eth1EventsChannel).setLatestPublishedDeposit(UInt64.valueOf(lastReplayedDepositIndex));
     inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
@@ -359,39 +373,6 @@ class Eth1DepositManagerTest {
     assertNoUncaughtExceptions();
   }
 
-  @Test
-  void shouldRetryWhenEth1ChainIsLessThanEth1FollowDistance() {
-    // Initial latest block number is less than ETH1_FOLLOW_DISTANCE
-    final BigInteger firstEth1BlockNumber = BigInteger.valueOf(60);
-    final Block firstLatestBlock = block(firstEth1BlockNumber, 100000);
-
-    // Second latest block number is greater than ETH1_FOLLOW_DISTANCE
-    final BigInteger secondEth1BlockNumber = BigInteger.valueOf(101);
-    final Block secondLatestBlock = block(secondEth1BlockNumber, 150000);
-
-    when(eth1Provider.getGuaranteedLatestEth1Block())
-        .thenReturn(SafeFuture.completedFuture(firstLatestBlock));
-
-    when(eth1Provider.getLatestEth1Block())
-        .thenReturn(SafeFuture.completedFuture(secondLatestBlock));
-
-    Constants.ETH1_FOLLOW_DISTANCE = UInt64.valueOf(100);
-    when(eth1DepositStorageChannel.replayDepositEvents()).thenReturn(NOTHING_REPLAYED);
-
-    manager.start();
-
-    inOrder.verify(eth1DepositStorageChannel).replayDepositEvents();
-    verify(eth1Provider).getGuaranteedLatestEth1Block();
-    verify(eth1Provider, never()).getGuaranteedEth1Block((UInt64) any());
-
-    assertThat(asyncRunner.hasDelayedActions()).isTrue();
-    asyncRunner.executeQueuedActions();
-
-    verify(eth1Provider, times(1)).getLatestEth1Block();
-    verify(eth1Provider).getGuaranteedEth1Block((UInt64) any());
-    assertNoUncaughtExceptions();
-  }
-
   private void withMinGenesisBlock(
       final BigInteger headBlockNumber, final BigInteger minGenesisBlockNumber) {
     final Block minGenesisBlock = block(minGenesisBlockNumber, MIN_GENESIS_BLOCK_TIMESTAMP);
@@ -400,22 +381,23 @@ class Eth1DepositManagerTest {
         .thenReturn(SafeFuture.completedFuture(minGenesisBlock));
   }
 
-  private void withFollowDistanceHead(final BigInteger number, final long timestamp) {
-    final Block latestBlock =
-        block(number.add(Constants.ETH1_FOLLOW_DISTANCE.bigIntegerValue()), timestamp + 100000);
-    final Block followDistanceHead = block(number, timestamp);
-    when(eth1Provider.getGuaranteedLatestEth1Block())
-        .thenReturn(SafeFuture.completedFuture(latestBlock));
-    when(eth1Provider.getGuaranteedEth1Block(UInt64.valueOf(number)))
-        .thenReturn(SafeFuture.completedFuture(followDistanceHead));
-  }
-
   private Block block(final BigInteger number, final long timestamp) {
     final Block block = mock(Block.class);
     when(block.getNumber()).thenReturn(number);
     when(block.getTimestamp()).thenReturn(BigInteger.valueOf(timestamp));
     when(block.getHash()).thenReturn(Bytes32.ZERO.toHexString());
     return block;
+  }
+
+  private void notifyHeadBlock(final BigInteger blockNumber, final long timestamp) {
+    final Block latestBlock = block(blockNumber, timestamp);
+    when(eth1Provider.getGuaranteedEth1Block(UInt64.valueOf(blockNumber)))
+        .thenReturn(SafeFuture.completedFuture(latestBlock));
+
+    final ArgumentCaptor<HeadUpdatedSubscriber> captor =
+        ArgumentCaptor.forClass(HeadUpdatedSubscriber.class);
+    verify(eth1HeadTracker, atLeastOnce()).subscribe(captor.capture());
+    captor.getValue().onHeadUpdated(UInt64.valueOf(blockNumber));
   }
 
   private void assertNoUncaughtExceptions() {

--- a/services/powchain/src/main/java/tech/pegasys/teku/services/powchain/PowchainService.java
+++ b/services/powchain/src/main/java/tech/pegasys/teku/services/powchain/PowchainService.java
@@ -146,7 +146,8 @@ public class PowchainService extends Service {
             eth1DepositStorageChannel,
             depositProcessingController,
             new MinimumGenesisTimeBlockFinder(eth1Provider, eth1DepositContractDeployBlock),
-            eth1DepositContractDeployBlock);
+            eth1DepositContractDeployBlock,
+            headTracker);
 
     eth1ProviderMonitor = new Eth1ProviderMonitor(eth1ProviderSelector, asyncRunner);
   }

